core/server/src/main/java/alluxio/master/file/meta/Inode.java

```diff
@@ -277,27 +296,100 @@ public T setOwner(String owner) {
   protected abstract T getThis();
 
   /**
-   * Acquires the read lock for this inode.
+   * Obtains a read lock on the inode. This call should only be used when locking the root or an
+   * inode by id and not path or parent.
    */
   public void lockRead() {
     mLock.readLock().lock();
   }
 
   /**
-   * Releases the read lock for this inode.
-   */
-  public void unlockRead() {
-    mLock.readLock().unlock();
+   * Obtains a read lock on the inode. Afterward, checks the inode state to ensure the parent is
+   * consistent with what the caller is expecting. If the state is inconsistent, an exception
+   * will be thrown and the lock will be released.
+   *
+   * @param parent the expected parent inode
+   * @throws InvalidPathException if the parent is not as expected
+   */
+  public void lockReadAndCheckParent(Inode parent) throws InvalidPathException {
+    lockRead();
+    if (mParentId != InodeTree.NO_PARENT && mParentId != parent.getId()) {
+      unlockRead();
+      throw new InvalidPathException(ExceptionMessage.PATH_INVALID_CONCURRENT_RENAME.getMessage());
+    }
   }
 
   /**
-   * Acquires the write lock for this inode.
+   * Obtains a read lock on the inode. Afterward, checks the inode state to ensure the full inode
+   * path is consistent with what the caller is expecting. If the state is inconsistent, an
+   * exception will be thrown and the lock will be released.
+   *
+   * This method assumes that the inode path to the parent has been read locked.
+   *
+   * @param parent the expected parent inode
+   * @param name the expected name of the inode to be locked
+   * @throws InvalidPathException if the parent and/or name is not as expected
+   */
+  public void lockReadAndCheckFullPath(Inode parent, String name) throws InvalidPathException {
+    lockReadAndCheckParent(parent);
+    if (!mName.equals(name)) {
```

Comment:
I called it check full path because with all the implicit conditions, checking the name and parent ensures the entire path is consistent. However I can see it can be confusing and I'm open to changing the name to something more descriptive of the actual operation.